platforms/nuttx/src/px4/common/px4_mtd.cpp

```diff
@@ -0,0 +1,458 @@
+
+	if (ret != OK) {
+		// FIXME: From the previous warning call, it looked like this should have been fatal error instead. Tried
+		// that but setting the bus speed does fail all the time. Which was then exiting and the board would
+		// not run correctly. So changed to PX4_WARN.
+		PX4_WARN("failed to set bus speed");
+	}
+
+	return 0;
+#endif
+}
+
+
+static int at24xxx_attach(mtd_instance_s &instance)
+{
+#if !defined(PX4_I2C_BUS_MTD)
+	PX4_ERR("Misconfiguration PX4_I2C_BUS_MTD not set");
+	return -ENXIO;
+#else
+
+	struct i2c_master_s *i2c = px4_i2cbus_initialize(PX4_I2C_DEVID_BUS(instance.devid));
+
+	if (i2c == nullptr) {
+		PX4_ERR("failed to locate I2C bus");
+		return -ENXIO;
+	}
+
+	/* start the MTD driver, attempt 5 times */
+	for (int i = 0; i < 5; i++) {
+		instance.mtd_dev = px4_at24c_initialize(i2c, PX4_I2C_DEVID_ADDR(instance.devid));
+
+		if (instance.mtd_dev) {
+			/* abort on first valid result */
+			if (i > 0) {
+				PX4_WARN("EEPROM needed %d attempts to attach", i + 1);
+			}
+
+			break;
+		}
+	}
+
+	/* if last attempt is still unsuccessful, abort */
+	if (instance.mtd_dev == nullptr) {
+		PX4_ERR("failed to initialize EEPROM driver");
+		return -EIO;
+	}
+
+	return 0;
+#endif
+}
+
+
+int px4_mtd_get_geometry(const mtd_instance_s *instance, unsigned long *blocksize, unsigned long *erasesize,
+			 unsigned long *neraseblocks,
+			 unsigned *blkpererase, unsigned *nblocks, unsigned *partsize)
+{
+	/* Get the geometry of the FLASH device */
+
+	FAR struct mtd_geometry_s geo;
+
+	int ret = instance->mtd_dev->ioctl(instance->mtd_dev, MTDIOC_GEOMETRY, (unsigned long)((uintptr_t)&geo));
+
+	if (ret < 0) {
+		PX4_ERR("mtd->ioctl failed: %d", ret);
+		return ret;
+	}
+
+	*blocksize = geo.blocksize;
+	*erasesize = geo.erasesize;
+	*neraseblocks = geo.neraseblocks;
+
+	/* Determine the size of each partition.  Make each partition an even
+	 * multiple of the erase block size (perhaps not using some space at the
+	 * end of the FLASH).
+	 */
+
+	*blkpererase = geo.erasesize / geo.blocksize;
+	*nblocks     = (geo.neraseblocks / instance->n_partitions_current) * *blkpererase;
+	*partsize    = *nblocks * geo.blocksize;
+
+	return ret;
+}
+
+/*
+  get partition size in bytes
+ */
+ssize_t px4_mtd_get_partition_size(const mtd_instance_s *instance, const char *partname)
+{
+	unsigned long blocksize, erasesize, neraseblocks;
+	unsigned blkpererase, nblocks, partsize = 0;
+
+	int ret = px4_mtd_get_geometry(instance, &blocksize, &erasesize, &neraseblocks, &blkpererase, &nblocks, &partsize);
+
+	if (ret != OK) {
+		PX4_ERR("Failed to get geometry");
+		return 0;
+	}
+
+	unsigned partn = 0;
+
+	for (unsigned n = 0; n < instance->n_partitions_current; n++) {
+		if (instance->partition_names[n] != nullptr &&
+		    partname != nullptr &&
+		    strcmp(instance->partition_names[n], partname) == 0) {
+			partn = n;
+			break;
+		}
+	}
+
+	return instance->partition_block_counts[partn] * erasesize;
```

Comment:
But it you decided to use `instances[i].partition_block_counts[part]` as number of blocks (not the erase blocks), do not forget that `px4_mtd_get_partition_size` here should be fixed:

`return instance.partition_block_counts[partn] * blocksize;`

Reply:
@lukegluke - Fixed and force pushed - Thank you